### PR TITLE
strengthen the Cardano ThreadNet test

### PR DESCRIPTION
**Pull request by nfrisby**
This is a lightweight addition to the test so that it fails if:

  (1) It progresses to Shelley even though the adopted update proposal didn't increment the major protocol version.

  (2) It doesn't progress to Shelley even though the update proposal incremented the major version and there were enough slots that a Shelley block should have been created.

### CODE REVIEW

File: ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs [hunk 1]
```diff
@@ -203,13 +237,58 @@ prop_simple_cardano_convergence TestSetup
...
+        k :: Word64
+        k = maxRollbacks setupK
+
+        w :: Word64
+        w = Shelley.computeStabilityWindow
+            setupK
+            (SL.sgActiveSlotCoeff genesisShelley)
+
+        isShelley :: HF.HardForkBlock '[ByronBlock, ShelleyBlock c] -> Bool
+        isShelley (HF.HardForkBlock (HF.OneEraBlock ns)) = index_NS ns == 1
```
Comment (mrBliss) [inline ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs:282]: You can use the `BlockShelley` pattern synonym from `O.C.Cardano.Block`, that's much prettier:
```haskell
isShelley :: CardanoBlock scc -> Bool
isShelley (BlockShelley _) = True
isShelley _                = False
```

File: ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs [hunk 2]
```diff
@@ -203,13 +237,58 @@ prop_simple_cardano_convergence TestSetup
...
+      , rsShelleySlots =
+          assert (w >= k) $
+          BoolProps.requiredIf $ t > w + 1 - k
+            -- logic: if we are ensured at least @k@ blocks in @w@ slots, then
+            -- we're ensured at least @1@ block in @w+1-k@ slots
+      }
+      where
+        TestConfig{numSlots} = setupTestConfig
+        NumSlots t           = numSlots
+        TestOutput{testOutputNodes} = testOutput
```
Comment (mrBliss) [inline ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs:271]: All or nothing :slightly_smiling_face:

File: ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs [hunk 3]
```diff
@@ -203,13 +237,58 @@ prop_simple_cardano_convergence TestSetup
...
     epochSizeShelley = sgEpochLength genesisShelley
 
+    -- the protocol version of the Byron era proposal
+    propPV :: CC.Update.ProtocolVersion
+    propPV =
+      if setupHardFork
+      then CC.Update.ProtocolVersion shelleyMajorVersion 0 0
+      else CC.Update.ProtocolVersion
+             byronMajorVersion
+             (byronInitialMinorVersion + 1)
```
Comment (mrBliss) [inline ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs:247]: I don't understand this, can you explain?

File: ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs (hunk 1 above)
Comment (nfrisby) [reply ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs]: Done! Thanks, much nicer.

File: ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs (hunk 2 above)
Comment (nfrisby) [reply ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs]: Ugh, copy paste.

File: ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs (hunk 3 above)
Comment (nfrisby) [reply ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs]: I've added a couple comments. Also see the comment on `TestSetup` and `setupHardFork`.

Does that help?

File: ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs [hunk 4]
```diff
@@ -203,13 +235,62 @@ prop_simple_cardano_convergence TestSetup
...
+
+        k :: Word64
+        k = maxRollbacks setupK
+
+        w :: Word64
+        w = Shelley.computeStabilityWindow
+            setupK
+            (SL.sgActiveSlotCoeff genesisShelley)
+
+        isShelley :: HF.HardForkBlock '[ByronBlock, ShelleyBlock c] -> Bool
```
Comment (mrBliss) [inline ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs:281]:
```suggestion
        isShelley :: CardanoBlock sc -> Bool
```

File: ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs [hunk 5]
```diff
@@ -203,13 +235,62 @@ prop_simple_cardano_convergence TestSetup
...
+    propPV :: CC.Update.ProtocolVersion
+    propPV =
+      if setupHardFork
+      then
+        -- this new version must induce the hard fork if accepted
+        CC.Update.ProtocolVersion shelleyMajorVersion 0 0
+      else
+        -- this new version must not induce the hard fork if accepted
+        CC.Update.ProtocolVersion
+          byronMajorVersion (byronInitialMinorVersion + 1) 0
```
Comment (mrBliss) [inline ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs:248]: Ah, so it is to test that proposals that should not trigger the transition to Shelley indeed don't, i.e., the goal of this PR.

File: ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs (hunk 3 above)
Comment (mrBliss) [reply ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs]: Was meant as a response: https://github.com/input-output-hk/ouroboros-network/pull/2282#discussion_r442243653

**Pull request merged**